cmd/bosun/expr/funcs.go

```diff
@@ -787,3 +813,225 @@ func Transpose(e *State, T miniprofiler.Timer, d *Results, gp string) (*Results,
+	if err != nil {
+		return nil, err
+	}
+
+	//service = service.Debug(true).Pretty(true)
+	ts := elastic.NewDateHistogramAggregation().Field("@timestamp").Interval(strings.Replace(interval, "M", "n", -1))
+	ds, err := opentsdb.ParseDuration(interval)
+	if err != nil {
+		return nil, err
+	}
+	if keystring == "" {
+		s = s.Aggregation("ts", ts)
+		data, err := json.MarshalIndent(s.Source(), "", "\t")
+		if err != nil {
+			panic("error generating raw elastic query")
+		}
+		fmt.Println(string(data))
+		result, err := service.SearchSource(s).Do()
+		if err != nil {
+			return nil, err
+		}
+		ts, found := result.Aggregations.DateHistogram("ts")
+		if !found {
+			return nil, fmt.Errorf("expected time series not found in elastic reply")
+		}
+		series := make(Series)
+		for _, v := range ts.Buckets {
+			series[time.Unix(v.Key/1000, 0).UTC()] = float64(v.DocCount) / ds.Seconds()
+		}
+		r.Results = append(r.Results, &Result{
+			Value: series,
+			Group: make(opentsdb.TagSet),
+		})
+		return r, nil
+	}
+	aggregation := elastic.NewTermsAggregation().Field(keys[len(keys)-1].Key)
+	aggregation = aggregation.SubAggregation("ts", ts)
+	if len(keys) > 1 {
+		for i := len(keys) - 2; i > -1; i-- {
+			aggregation = elastic.NewTermsAggregation().Field(keys[i].Key).SubAggregation("g_"+keys[i+1].Key, aggregation)
+		}
+	}
+	s = s.Aggregation("g_"+keys[0].Key, aggregation)
+	data, err := json.MarshalIndent(s.Source(), "", "\t")
+	fmt.Println(string(data))
+	if err != nil {
+		panic("error generating raw elastic query")
+	}
+	fmt.Println(string(data))
+	result, err := service.SearchSource(s).Do()
+	if err != nil {
+		return nil, err
+	}
+	top, ok := result.Aggregations.Terms("g_" + keys[0].Key)
+	if !ok {
+		return nil, fmt.Errorf("top key g_%v not found in result", keys[0].Key)
+	}
+	for _, b := range top.Buckets {
+		tags := make(opentsdb.TagSet)
+		key := fmt.Sprintf("%v", b.Key)
+		if keys[0].Pattern != nil && !keys[0].Pattern.MatchString(key) {
+			continue
+		}
+		tags[keys[0].Key] = key
+		desc(b, tags, 1, keys, r, e, ds)
+	}
+	return r, nil
+}
+
+func desc(b *elastic.AggregationBucketKeyItem, tags opentsdb.TagSet, depth int, keys []keyMatch, r *Results, e *State, ds opentsdb.Duration) {
+	if ts, found := b.DateHistogram("ts"); found {
+		if e.squelched(tags) {
+			return
+		}
+		series := make(Series)
+		for _, v := range ts.Buckets {
+			series[time.Unix(v.Key/1000, 0).UTC()] = float64(v.DocCount) / ds.Seconds()
+		}
+		r.Results = append(r.Results, &Result{
+			Value: series,
+			Group: tags.Copy(),
+		})
+		return
+	}
+	n, ok := b.Aggregations.Terms("g_" + keys[depth].Key)
+	if !ok {
+		panic(fmt.Sprintf("%v not found at depth %v", "g_"+keys[depth].Key, strconv.Itoa(depth)))
+	}
+	for _, item := range n.Buckets {
+		key := fmt.Sprintf("%v", item.Key)
+		if keys[depth].Pattern != nil && !keys[depth].Pattern.MatchString(key) {
+			continue
+		}
+		tags[keys[depth].Key] = key
+		desc(item, tags, depth+1, keys, r, e, ds)
```

Comment:
Need to create a new tags here, not modify in place. This is a depth-first search, so after the first one tags will be fully populated and you will be only changing certain nodes.